### DISCUSSION TITLE
Accessibility: Do we have it in mind?

### DISCUSSION THREAD

**Discussion by saona-raimundo** [kas-gui/kas#263] [General]
Hi there!

Thank you for the great work!

I wanted to ask for accessibility features in the kas gui system.

For example:
- Screen-reader integration
- Localized zoom
- Keyboard control
- Tab navigation
- etc...

Btw, I have to applaud the Stepless DPI scaling! :D

## Details

I consider accessibility part of "empowerment", as in the [rust book foreword](https://doc.rust-lang.org/book/foreword.html). 

Other GUI frameworks in Rust have not achieved accessibility support ([egui](https://github.com/emilk/egui/issues/167), [druid](https://github.com/linebender/druid/issues/299), dear imgui ([link ot C lib](https://github.com/ocornut/imgui/issues/4122)), [iced](https://github.com/iced-rs/iced/issues/552), [conrod](https://github.com/PistonDevelopers/conrod/issues/1040), [azul](https://github.com/fschutt/azul/issues/108)). Also, accessibility is much easier to have included in an early stage of a project than an after-thought. 

100% accessibility is not the goal, but it is important to make a conscious effort for accessibility. So, in the kas gui system, is accessibility in mind?

## References

Most of these references are about accessibility in the web, but they apply just as much to desktop apps.
- [Accessibility handbook](https://accessibility.oit.ncsu.edu/it-accessibility-at-nc-state/developers/accessibility-handbook/)
- [Web Content Accessibility Guidelines](https://www.w3.org/WAI/standards-guidelines/wcag/)
- [Accessibility - Learn web development | MDN](https://developer.mozilla.org/en-US/docs/Learn/Accessibility)
- [Reakit](https://reakit.io/): Build accessible rich web apps with React
- [Mui](https://mui.com/) (Another React accessibility framework)

**Comment by dhardy** [comment]
Hello,

> Screen-reader integration

No current plans here. I see some discussion related to this in https://github.com/emilk/egui/issues/167.

> Localized zoom

How is this feature supposed to work? I see you wrote https://github.com/emilk/egui/issues/941. The fundamental questions here are (1) how is zoom controlled and (2) which part of the screen is zoomed.

- Zoom an area around the mouse: there are already desktop plugins (at least on the desktop I use, KDE) to do this.
- Enlarge the whole application at start-up: this is already possible, either by changing the screen scale factor under display settings, or (on `X11`) via environment variable, e.g. `export WINIT_X11_SCALE_FACTOR=1.5`. See [How is the scale factor calculated](https://docs.rs/winit/0.25.0/winit/dpi/index.html#how-is-the-scale-factor-calculated).
- Zoom some region of the application relative to everything else, e.g. a "zoomable frame". This would need (1) controls and (2) some way of creating a sub-region with different scale factor. Not too hard I think.

> Keyboard control

What were you looking for that doesn't already exist?

> Tab navigation

This should work; if not open a bug report. See e.g. https://github.com/kas-gui/kas/issues/231 (but open a new report for new issues).

> So, in the kas gui system, is accessibility in mind?

Keyboard control has been in mind from early on. Screen readers no, mainly since I lack specific experience/motivation.

---

By the way, why is this issue created against the tutorials repo? Should I move it?

**Reply by saona-raimundo** [reply]
Thank you for the complete reply!

I was just starting with kas and thought that asking was the right thing to do after reading the tutorials. It is not like I have a feature request, or maybe adding some docs in the tutorial? We could add tab-navigation and keyboard control, i think they are good selling points!

I just found this repo about [accessibility in UIs](https://github.com/AccessKit/accesskit), which solves many of my questions (it is WIP). When ready, screen-reader integration should be "for free"!

----

About the correct repo, yes, sorry, I meant to put it in kas. Do you mind moving it?

**Reply by dhardy** [reply]
If you are talking about revising the README, please feel free to make an edit/PR.

**Comment by dhardy** [comment]
> We could add tab-navigation and keyboard control, i think they are good selling points!

What do you mean by this? Add some documentation about it here?

Tab-navigation is enabled by default (though it can be broken if a widget handles `Command::Tab` — several real apps like text editors do basically this).

Keyboard control: the calculator tutorial does mention key bindings. Other types of keyboard controls are not supported yet (nor are context menus, which bear some resemblance to how global shortcuts would work).

**Reply by saona-raimundo** [reply]
> Tab-navigation is enabled by default

Do you think it is worth mentioning it in [Hello: push button](https://kas-gui.github.io/tutorials/hello.html) example? :)
There is a button there and it can be accessed by tabbing.

> Keyboard control: the calculator tutorial does mention key bindings. 

It does, indeed! Although a little late in the tutorial, keyboard control is there. 

I was playing with it and had the problem that it "did not work for composed key-bindings": 
- It works perfectly for single keys: Typing `Alt + 9` is like pushing the button `9`.
- It does not work for composed key-bindings: In my keyboard, to type `+`, I have to type `Shift + =` (US keyboard). I tried typing `Alt + Shift + =`, but it did not work, it behaves as if I had pressed only `Alt + =` (no shift).

**Comment by dhardy** [comment]
FYI the latest version of Kas has (limited) support for screen readers (except input fields) and IME/virtual keyboards. These each have issue trackers.
